jmaps

```diff
@@ -72,25 +72,56 @@ for pid in $(pgrep -x java); do
 	mapfile=/tmp/perf-$pid.map
 	[[ -e $mapfile ]] && rm $mapfile
 
-	cmd="cd $AGENT_OUT; $JAVA_HOME/bin/java -Xms32m -Xmx128m -cp $AGENT_JAR:$JAVA_HOME/lib/tools.jar net.virtualvoid.perf.AttachOnce $pid $opts"
-	(( debug )) && echo $cmd
-
 	user=$(ps ho user -p $pid)
-	if [[ "$user" != root ]]; then
-		if [[ "$user" == [0-9]* ]]; then
-			# UID only, run sudo with #UID:
-			cmd="sudo -u '#'$user sh -c '$cmd'"
-		else
-			cmd="sudo -u $user sh -c '$cmd'"
+
+	container=$(cat /proc/$pid/cgroup | grep :/docker/ | awk -F/ '{ print $NF }' | head -n1)
+	if [[ $container != "" ]]; then
+		echo "Prorcess $pid is in container $container"
+
+		tmpfile=$(mktemp -u)
+
+		container_pid=$(cat /proc/$pid/status | grep ^NSpid | awk '{ print $NF }')
+		nsenter="sudo nsenter --target $pid --setuid $user --setgid $(sudo nsenter --target $pid --mount --pid id -g $user) --mount --pid"
+
+		java_home=$(sudo cat /proc/$pid/environ | xargs --null --max-args=1 | grep ^JAVA_HOME | awk -F= '{ print $2 }')
+		if [[ $java_home == "" ]]; then
+			java_home=$($nsenter jrunscript -e 'java.lang.System.out.println(java.lang.System.getProperty("java.home"));')
 		fi
-	fi
 
-	echo "Mapping PID $pid (user $user):"
-	if (( debug )); then
-		time eval $cmd
+		cat $AGENT_JAR | $nsenter dd status=none of=$tmpfile
```

Comment:
cat and dd? why not cp to the /proc/PID/root.

This needs error checking.

We should also think seriously about dropping files in people's containers just because someone/something ran jmap. Is there any way to avoid this? Eg, running java with chroot/nsenter for mounts. Or passing in the agent java as a virtual device (like stdin) for java to run, than just copying it to their file system.

Reply:
Didn't know about `/proc/PID/root`, that definitely makes it easier.

Regarding error checking: what do you have in mind? If copy fails, should we print an error and continue to the next loop iteration?

I thought about using `/dev/sdtin`, but I have two files and there's only one stdin. I'll look into chroot / another mount namespace.

Reply:
Another thing here is perf-map-agent also creates files in container's `/tmp`.